modeshape-jcr/src/main/java/org/modeshape/jcr/spi/index/provider/IndexChangeAdapters.java

```diff
@@ -334,26 +439,35 @@ protected void removeNode( String workspaceName,
                                    Path path,
                                    Name primaryType,
                                    Set<Name> mixinTypes ) {
-            if (includeRoot || path.isRoot()) {
-                index.remove(nodeKey(key));
-            }
+            index().remove(nodeKey(key));
         }
 
         @Override
-        protected void completeChanges() {
-            index.commit();
-            super.completeChanges();
-        }
-
-        @Override
-        protected void completeWorkspaceChanges() {
-            index.commit();
-            super.completeWorkspaceChanges();
+        protected void renameNode( String workspaceName, 
+                                   NodeKey key, 
+                                   Path newPath, 
+                                   Path.Segment oldSegment, 
+                                   Name primaryType,
+                                   Set<Name> mixinTypes ) {
+            PathFactory pathFactory = context.getValueFactories().getPathFactory();
+            Path oldPath = pathFactory.create(newPath.subpath(0, newPath.size()), oldSegment);
+            String nodeKey = nodeKey(key);
+            index().remove(nodeKey, valueOf(oldPath));
+            index().add(nodeKey, valueOf(newPath));
         }
 
         @Override
-        public String toString() {
-            return getClass().getSimpleName() + "(\"" + index.getName() + "\")";
+        protected void reorderNode( String workspaceName, 
+                                    NodeKey key, 
+                                    Name primaryType, 
+                                    Set<Name> mixinTypes, 
+                                    NodeKey parent,
+                                    Path newPath, 
+                                    Path oldPath, 
+                                    Path reorderedBeforePath ) {
+            String nodeKey = nodeKey(key);
+            index().remove(nodeKey, valueOf(oldPath));
+            index().add(nodeKey, valueOf(newPath));
```

Comment:
well, I guess it's not as bad as I though (ran a quick SNS reordering test). The only actual changes are between the source node and the target node, since SNSs are computed each time (not stored).
So given `node[1], node[2], node[3], node[4]` performing a `reorder(node[4],node[1])` will still yield the same result, but the actual keys which have changed positions are `node[1]` and `node[4]`. 
So there is still some work to update the `reorder` events, but it's doable (unlike my previous "fear")